zpay32/invoice.go

```diff
@@ -1012,3 +1026,46 @@ func uint64ToBase32(num uint64) []byte {
 	// We only return non-zero leading groups.
 	return arr[i:]
 }
+
+// LtcToBtcParams applies the relevant chain configuration parameters that
+// differ for litecoin similar to chainparams.applyLiteCoinParams(). This
+// function is used in place of an interface to abstract over which chain
+// the parameters are for.
+func LtcToBtcParams(params *litecoinCfg.Params) *chaincfg.Params {
```

Comment:
Perhaps I misunderstand something, but doesn’t [chainparams.go](https://github.com/lightningnetwork/lnd/blob/master/chainparams.go#L60) already do this?

Reply:
Yes pretty much, but it's a private method in chainparams.go and it doesn't return a new instance of btcd/chainCfg.Params, so I copied that functionality over as I mentioned in the comments. Eventually I figure both of those methods will go away.

Reply:
Maybe it is better to use/expose `activeNetParams` [which is set here](https://github.com/lightningnetwork/lnd/blob/master/config.go#L302)  from the main package(since it will have params applied and won’t create any code redundancy). IMO logic related to the chain shouldn’t leak too much into all the packages, but I have no idea what the overhaul would do, since this applying is a “temporary hack”(as noted in comments).

Reply:
Yeah I noted a todo for a generic chain param interface as per halseth's comment, I'd be removing this method as part of that change.

Reply:
If you can reuse the code from chainparams, that would be preferable! Maybe `activeNetParams` can be passed to the package?

Reply:
@halseth I've been thinking about how to approach this, wouldn't relying on `activeNetParams` conflict if you're decoding an invoice for a chain other than the active one? I think I ran into a similar problem with my first attempt to work this by with reusing the `applyLitecoinParams` method, where I would accidentally overwrite parameters in activeNetParams when I decoded a litecoin invoice. 

The `applyLitecoinParams` method in `chainparams` currently just overwrites parameters with the litecoin testnet parameters. So I figure I'd probably have to change that to accept a parameter for whether it should apply mainnet, testnet, or simnet parameters. Then I'd have to call it in a way that doesn't override active params. Let me know what you think, thanks.

Reply:
I'm leaning towards letting `Decode` take a `chaincfg` (or equivalent) argument, and fail decoding the invoice if it's meant for an unknown network. This should simplify the logic a fair bit. In the future, when it should support multiple chains, we can let it take a list of known networks.

Reply:
To me it seems a bit cumbersome to have to pass in a network parameter to `decodepayreq` for instance. Also I figure it might be possible once multiple chains are supported that a desired use case is to decode an invoice without necessarily knowing what network it's intended for beforehand. Perhaps another approach is to fail the decoding if the decoded invoice does not match any of the active networks?

Reply:
The problem with decoding without passing in the known networks, is that you need to add a new entry (possibly more: mainnet, simnet, testnet, regtest etc) to the switch statement for each new network. Having it just check again the passed known ones would be simpler. (And safer? Making sure the software is able to understand the invoice)

Checking against the active networks would be similar, but is not really something we can easily do at the moment (since it is assumed only one network is active). Passing in known network would let us pass in the one active network for now, and just pass in the active networks in the future.

Reply:
Right, the switch statement would have to grow for each new supported network. I would not expect safety to be an issue though, as it fails unknown networks in any case. But yeah, I suppose it could take `activeNetParams` like you suggested originally, and then when multiple active chains are supported it could be expanded to take all active networks and check if any match the invoice.